lib/SILOptimizer/Transforms/StringSwitch.cpp

```diff
@@ -0,0 +1,258 @@
+
+// Optimize multiple string literal comparisons against a single String type.
+// This can be used to optimize a series of string comparison ifs
+// or, more commonly, a string-switch-statement.
+class StringSwitchPass : public SILFunctionTransform {
+  bool doOneIteration() {
+    SILFunction *F = getFunction();
+    SILModule &module = F->getModule();
+
+    DominanceAnalysis *domAnalysis = PM->getAnalysis<DominanceAnalysis>();
+    DominanceInfo *domInfo = domAnalysis->get(F);
+    DominanceOrder domOrder(&F->front(), domInfo);
+
+    // Find all the string literals used in a series of comparisons agains
+    // one "normal" string.
+    SmallVector<StringLiteralInst *, 12> stringLiterals;
+    SmallVector<ApplyInst *, 12> stringCompares;
+    SILValue subject;
+    // Iterate through all the blocks in the function using dominance order
+    // so that we don't miss the first comparison because we started with the
+    // seconcd comparison (hypothetically).
+    while (auto *block = domOrder.getNext()) {
+      domOrder.pushChildren(block);
+
+      for (auto &inst : *block) {
+        auto *apply = dyn_cast<ApplyInst>(&inst);
+        if (!apply || !apply->getReferencedFunctionOrNull())
+          continue;
+        // Filter out the string.equals functions.
+        if (!apply->getReferencedFunctionOrNull()->hasSemanticsAttr(
+                semantics::STRING_EQUALS))
+          continue;
+
+        // Make sure that we are comparing agains the same subject as the
+        // other comparisons.
+        bool argsNotEqual = apply->getArgument(1) != subject &&
+                            apply->getArgument(0) != subject;
+
+        // If we've already collected one comparison, and nither of the
+        // apply args match the subject, or the first string literal isn't
+        // dominated by the comparison call, then continue.
+        if (!stringLiterals.empty() &&
+            (argsNotEqual ||
+             !domAnalysis->get(F)->dominates(stringLiterals.front(), apply)))
+          continue;
+
+        unsigned subjectArgIndex = 1;
+        auto info = StringLiteralInfo::create(apply->getArgument(0));
+        if (!info) {
+          info = StringLiteralInfo::create(apply->getArgument(1));
+          subjectArgIndex = 0;
+        }
+        if (!info)
+          continue;
+        stringLiterals.push_back(info.getValue().inst);
+        stringCompares.push_back(apply);
+        subject = apply->getArgument(subjectArgIndex);
+      }
+    }
+
+    if (stringLiterals.empty())
+      return false;
+
+    // Set the insert point after the first string compare instructions
+    // (garenteed to be after the first string literal instruction).
+    SILBuilder builder(stringCompares.front());
+
+    SmallVector<ValueDecl *, 1> results;
+    // We'll let ObjectOutliner::replaceFindStringCall do the heavy lifting.
+    builder.getASTContext().lookupInSwiftModule("_findStringSwitchCase",
```

Comment:
`_findStringSwitchCase` only exists if there is a string enum in the source file. I suspect that `StringSwitch.swift` isn't included unless there is a string enum. Does anyone know how I should fix that?